### PR TITLE
Separate constants from DocumentDB into DocsApiConstants

### PR DESCRIPTION
**Pull request by EricBorczuk**
**What this PR does**:
This PR refactors a bunch of the constants from DocumentDB into a combination of DocsApiConstants (Renamed from QueryConstants) and DocsApiConfiguration. As a result of this, a few things have changed:
1) In a bunch of places, DocsApiConfiguration or its methods gets used instead of DocumentDB
2) `allColumns` and its derivatives were consolidated into one place. Before, there were two separate places for generating the columns for writing and reading, not anymore - but this also means that the column order expectation was changed for reading (in other words, writes require a certain order of columns for binding params, while selecting didn't, but now they are both the same)
3) newBindMap is no longer a part of DocumentDB

The only thing missing from this PR w.r.t the issue description is an Immutable class that caches calls to `allColumns` and its sibling functions; I could add that in this PR but it was getting kind of large and I think it's reasonable to review in this state.

**Which issue(s) this PR fixes**:
Fixes https://github.com/stargate/stargate/issues/1235

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated

### CODE REVIEW

**Comment by ivansenic** [comment]
@EricBorczuk The changes are rather large and include changes of the signatures in the search query builder, can we discuss first what s being done and why, it would be easier for me to review this monster?

**Review by ivansenic — CHANGES_REQUESTED** [review]
See my comment above..

**Comment by tatu-at-datastax** [comment]
It would be great to get this (and the other related refactoring) merged -- of course considering open issues etc -- since V2 prototype work will be focused on `restapi` module, and unfortunately Docs API is bundled with it.
Current thinking is that v2 work will happen in a branch, requiring steady merging from `master`. That's fine for fixes, feature changes, but can sometimes be more hassle with bigger refactorings.

**Comment by EricBorczuk** [comment]
@tatu-at-datastax Agreed - for this PR, I just incorporated the requested changes; I believe tests will pass in its current state. Its base https://github.com/stargate/stargate/pull/1259 is actually ready to merge after I get an approval. So I think I can get these out in the next day or so!

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java:259]: I was thinking that such things we can extract into immutable class as I specified in the #1235. Since, this is too big maybe we can keep that issue open and just do that on top of this PR. Wdyt?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsApiConfiguration.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsApiConfiguration.java:10]: why was this moved to static field instead of the function? the idea was that with the function you can override it in tests with a mocked value. This way we can not?

I would also be fine if you move this to the constants class?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsApiConfiguration.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsApiConfiguration.java:25]: should we extract those to the fields and return the field here? this way we are always parsing the strings, but sure why would we have that overhead?

Also fine to move to constants or some properties class

File: restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/ReactiveDocumentResourceV2.java:304]: I would still keep the static fields in the constants and use for this purpose.

File: restapi/src/main/java/io/stargate/web/docsapi/resources/CollectionsResource.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/CollectionsResource.java:308]: inject instance. not default please

File: restapi/src/main/java/io/stargate/web/docsapi/exception/ErrorCode.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/exception/ErrorCode.java:60]: revert to method?

File: restapi/src/main/java/io/stargate/web/docsapi/service/query/DocsApiConstants.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/query/DocsApiConstants.java:26]: OK now we need to add Java doc here for everything so that readers can understand what each value or function does!

File: restapi/src/main/java/io/stargate/web/docsapi/service/util/DocsApiUtils.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/util/DocsApiUtils.java:60]: do you think we should provide alternatives to the extended methods here that take the configuration as parameter?

```
convertArrayPath(String path, DocsApiConfiguraiton config) {
  return convertArrayPath(path, config.get..);
}
```

File: restapi/src/main/java/io/stargate/web/resources/Db.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/resources/Db.java:97]: inject, not static

**Review by ivansenic — CHANGES_REQUESTED** [review]
This look really good now.. Tests are failing so I set this to request changes.. I was lazy to look to the tests atm as they are not green, but I assume tests will not be a problem.. 

I left few comments as I was going through, I would like to know what do you think about that as well..

**Comment by ivansenic** [comment]
> It would be great to get this (and the other related refactoring) merged -- of course considering open issues etc -- since V2 prototype work will be focused on `restapi` module, and unfortunately Docs API is bundled with it. Current thinking is that v2 work will happen in a branch, requiring steady merging from `master`. That's fine for fixes, feature changes, but can sometimes be more hassle with bigger refactorings.

@tatu-at-datastax are you sure V2 will tackle both rest and docs, I know they in the same module now, but I guess they will be completely separated in V2?

**Comment by tatu-at-datastax** [comment]
@ivansenic We don't fully know details of V2 here wrt whether one service would handle both REST and Document API but ideally I think they would be separate. But even if they were left co-located I think separating them within module makes sense -- some parts can remain shared, as long as they are cleanly separate so that if we wanted to we could create separate "restdoc-common" package, if that makes sense.

It may get bit tricky with constants, but I would probably prefer RestApiConstants vs DocumentApiConstants for everything that can be separated.

File: restapi/src/test/java/io/stargate/web/docsapi/dao/DocumentDBTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/dao/DocumentDBTest.java:15]: not quite sure how this works now, but I think @tatu-at-datastax removed all direct guava usages and just depended on the shadowed version.. Would you be up using `Arrays.asList()` or `Collections.singeltonList()` just to avoid again the same old problem?

File: restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java:53]: For proper unit tests, we should use the mocked version, mock depending methods (I think this test has none) and then verify component is using values the mocked values..

File: restapi/src/test/java/io/stargate/web/docsapi/service/ReactiveDocumentServiceTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/ReactiveDocumentServiceTest.java:88]: Mocked version please..

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/ExpressionParserIntTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/ExpressionParserIntTest.java:63]: mocked version please, verify proper values if needed (not sure)

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/DocumentsSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/DocumentsSearchQueryBuilderTest.java:49]: mocked version, as we have in the other SearchQueryBuilderTests

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FilterExpressionSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FilterExpressionSearchQueryBuilderTest.java:52]: mocker version

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FilterPathSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FilterPathSearchQueryBuilderTest.java:50]: mocked version

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FullSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/FullSearchQueryBuilderTest.java:36]: not used?

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/PathSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/PathSearchQueryBuilderTest.java:42]: not used?

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/PopulateSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/PopulateSearchQueryBuilderTest.java:36]: not used?

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/SubDocumentSearchQueryBuilderTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/db/impl/SubDocumentSearchQueryBuilderTest.java:41]: not used?

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/BaseResolverTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/BaseResolverTest.java:59]: :rocket: :guitar:

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/filter/impl/InMemoryCandidatesFilterTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/filter/impl/InMemoryCandidatesFilterTest.java:126]: why is this changed now? any ideas?

File: restapi/src/test/java/io/stargate/web/docsapi/service/util/DocsApiUtilsTest.java
Comment (ivansenic) [inline restapi/src/test/java/io/stargate/web/docsapi/service/util/DocsApiUtilsTest.java:59]: can you just add one test that simulates array index higher that given max? and verify correct error being returned?

**Review by ivansenic — APPROVED** [review]
Approving as it's awesome.. Some minor testing fixes and additions needed.. Also rebase need due to the #1283

File: restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/filter/impl/InMemoryCandidatesFilterTest.java
Comment (EricBorczuk) [reply restapi/src/test/java/io/stargate/web/docsapi/service/query/search/resolver/filter/impl/InMemoryCandidatesFilterTest.java:126]: This is changed because now the search path and the write path use the same function for generating column names. The search path historically didn't care about the order of these columns, but for parameter binding purposes, the write side expects the columns to always be in the order `key, p0, ...pN, leaf, text_value, dbl_value, bool_value`

**Comment by ivansenic** [comment]
@EricBorczuk did you address the comments above before merging? if not I can create a quick followup?

**Pull request merged**
